add repositoryclient_mockup that implements repositoryclient interface with mock data

lynnsh opened the pull request:
- added GetBranch method that queries a specified branch using GitHub API
- refactored CommitInfo into Commit and Commit into CommitRef
- added RepositoryClient interface that implements Client interface with the addition of AddRepository function (used in testing, so that it is possible to add different repositories when testing with repositoryclient mockup; currently it's 1-1 relationship between repositories and organizations in repositoryclient mockup)
- modified test cases and repository_client to support and test with AddRepository

lynnsh commented on testing/repositoryclient_mockup.go:
```diff
@@ -0,0 +1,47 @@
+package vtesting
```
Done.

lynnsh commented on testing/repositoryclient_mockup.go:
```diff
@@ -0,0 +1,47 @@
+)
+
+type repositoryclient struct {
+}
```
Done. But since the Client interface does not support this method, I added another interface, RepoClient, with the only addition of this method. Is it the best solution?

catherinejones replied:
It's definitely not a bad solution! Alternatively, we could export `repositoryclient`, and allow other packages to interact with it directly!

catherinejones commented on repository/github/queries.go:
```diff
@@ -148,25 +148,25 @@ func createNewCommitInfo(queryResult *commitInfoQuery, checkSuites []checkSuite,
 
 // newDefaultBranchResult calls the defaultBranchQuery and populates the results with the respective variables
-func newDefaultBranchResult(ctx context.Context, ghc ghGraphQLClient, commitURL string) (repository.DefaultBranch, error) {
+func newDefaultBranchResult(ctx context.Context, ghc ghGraphQLClient, commitURL string) (repository.Branch, error) {
```
Do you think it would be difficult to generalize this function and add a separate `DefaultBranch` function which calls it?

lynnsh replied:
I created a new query - branchQuery that searches for the master branch. It gets called by GetBranch.
GetDefaultBranch queries for the default branch.

catherinejones replied:
Could you update the `branchQuery` to accept the branch name as a string? I think it'd be valuable to be able to limit the query to a specific branch if applicable

catherinejones commented on testing/repository.go:
```diff
@@ -0,0 +1,13 @@
+
+// Repository contains repository information pertaining to a repository
+type Repository struct {
+	org   repository.Organization
```
Since we're using this for testing, could we export these fields?

```suggestion
	Organization   repository.Organization
```

lynnsh replied:
Done. But I kept the variable name as Org, so that it is different from the variable type. Or should it be called Organization?

catherinejones commented on testing/repository_client.go:
```diff
@@ -0,0 +1,93 @@
+	"github.com/Shopify/voucher/repository"
+)
+const (
+    githubRegex = "^(?:https?|git)(?:://|@)(?P<vcs>[^/:]+)[/:](?P<org>[^/.]+)/(?P<name>[^/.]+)(?:.git)?"
```
Can we rename this to `repoRegex`

lynnsh replied:
Done

catherinejones commented on testing/repoclient.go:
```diff
@@ -0,0 +1,14 @@
+)
+
+// Client is a client for a version control source
+type RepoClient interface {
```
Could we name this RepositoryClient instead?

lynnsh replied:
Done

catherinejones commented on testing/repository_client.go:
```diff
@@ -0,0 +1,93 @@
+//GetCommitInfo creates commit info with provided
+func (c *repositoryclient) GetCommit(ctx context.Context, commitURL string) (repository.Commit, error) {
+    var commit repository.Commit
+    if len(c.repos) == 0 {
```
What do you think of removing this code and only using our new Repository map?

lynnsh replied:
Done

catherinejones commented on testing/repository_client.go:
```diff
@@ -0,0 +1,93 @@
+    		commit = repository.Commit{commitURL, nil, "bad", true, nil}
+    	}
+    } else {
+        re, _ := regexp.Compile(commitRegex)
```
Since we're ignoring the error, this might make us have an unusable regex! You can avoid that by using

```suggestion
        re := regexp.MustCompile(commitRegex)
```

`MustCompile` will panic if the regex is malformed, which will save us from unexpected bugs if someone changes the regex constant!

lynnsh replied:
Good to know! Done.

catherinejones commented on testing/repository_client.go:
```diff
@@ -0,0 +1,93 @@
+        org_name := re.FindStringSubmatch(commitURL)[2]
+        
+        branches := c.repos[org_name].branches
+        for i:=0; i < len(branches); i++ {
```
You can simplify this by running:

```suggestion
        for branchName, branch := range  c.repos[org_name].branches {
```

You can then interact with `branch.CommitRefs` directly!

lynnsh replied:
Done

catherinejones commented on testing/repository.go:
```diff
@@ -0,0 +1,13 @@
+	org   repository.Organization
+	name string
+	commits map[string]repository.Commit
+	branches []repository.Branch
```
Could we make this into a `map[string]repository.Branch` to speed up branch lookups?

lynnsh replied:
Done

lynnsh commented on repository/github/client.go:
```diff
@@ -58,20 +58,20 @@ func (ghc *client) GetOrganization(ctx context.Context, uri string) (repository.
 
 // GetDefaultBranch retrieves the necessary GitHub default branch information used in Voucher's checks
-func (ghc *client) GetDefaultBranch(ctx context.Context, commitURI string) (repository.DefaultBranch, error) {
+func (ghc *client) GetBranch(ctx context.Context, commitURI string) (repository.Branch, error) {
```
Done.

lynnsh commented on repository/github/client.go:
```diff
@@ -58,20 +58,20 @@ func (ghc *client) GetOrganization(ctx context.Context, uri string) (repository.
 // GetDefaultBranch retrieves the necessary GitHub default branch information used in Voucher's checks
-func (ghc *client) GetDefaultBranch(ctx context.Context, commitURI string) (repository.DefaultBranch, error) {
+func (ghc *client) GetBranch(ctx context.Context, commitURI string) (repository.Branch, error) {
 	defaultBranchResult, err := newDefaultBranchResult(ctx, ghc.ghClient, commitURI)
```
Done.

catherinejones commented on repository/github/branch_query.go:
```diff
@@ -0,0 +1,30 @@
+
+import "github.com/shurcooL/githubv4"
+
+// defaultBranchQuery is the GraphQL query for retrieving information pertaining to the repository's default branch
```
Could you update the comment here?

lynnsh replied:
Done

catherinejones commented on repository/github/queries_test.go:
```diff
@@ -600,13 +600,103 @@ func TestCreateNewDefaultBranch(t *testing.T) {
+			input: func() *branchQuery {
+				res := new(branchQuery)
+				res.Resource.Typename = "Commit"
+				//res.Resource.Commit.Repository.Ref.Name = "master"
```
Why is this commented? Is it something we should be using in our query?

lynnsh replied:
I checked how this test file works and forgot to uncomment that line. Fixed now.

catherinejones commented on repository/github/queries_test.go:
```diff
@@ -600,13 +600,103 @@ func TestCreateNewDefaultBranch(t *testing.T) {
+			}(),
+			mask: []string{"Resource.Typename", "Resource.Commit.Repository.Ref.Target.Commit.History.Nodes"},
+			expected: repository.Branch{
+				Name:    "",
```
Shouldn't this be the name of the branch we requested?

lynnsh replied:
Done

catherinejones commented on testing/repoclient.go:
```diff
@@ -0,0 +1,15 @@
+
+// RepositoryClient is a client for a version control source
+type RepositoryClient interface {
+	GetCommit(ctx context.Context, commitURL string) (r.Commit, error)
```
You can simplify this by changing it to:

```golang
type RepositoryClient interface {
     repository.Client
     AddRepository(org r.Organization, name string, commits map[string]r.Commit, branches map[string]r.Branch)
}
```

This will make `RepositoryClient` extend the `repository.Client` interface!

lynnsh replied:
Done.

catherinejones commented on testing/repository_client.go:
```diff
@@ -0,0 +1,85 @@
+}
+
+//NewClient creates a new repository client
+func NewClient() (RepositoryClient, error) {
```
`repositoryclient` looks great! 🎉

catherinejones commented on testing/repository_client.go:
```diff
@@ -0,0 +1,85 @@
+    
+    master := c.repos[org_name].Branches["master"]
+    commits := master.CommitRefs
+    for i:=0; i < len(commits); i++ {
```
Like the other functions, you can interate through slices using `range`:

```suggestion
    for i, commit := range branch.CommitRefs {
```

lynnsh replied:
Done

catherinejones commented on testing/repository_client.go:
```diff
@@ -0,0 +1,85 @@
+    re := regexp.MustCompile(repoRegex)
+    org_name := re.FindStringSubmatch(commit.URL)[2]
+    
+    master := c.repos[org_name].Branches["master"]
```
Do we always want to get the `master` branch? What if we're working on a different branch and that commit only exists there?

lynnsh replied:
Done (the name should be provided to the method to query for a specific branch)

lynnsh commented:
Changes done in the last 2 commits:
- added OrgCheck that gets build data from grafeas using MetadataClient and compares it to the data provided by GitHub
- updated repository.Client to accept BuildData instead of Commit
    - this created an import cycle: repository.Client requires voucher.BuildDetail and voucher.RepositoryCheck requires repository.Client
        - as a solution, BuildDetail and BuildArtifact were moved to repository package (not sure if it is the best solution)
- updated branch and org queries to search on repository url rather than a commit url (this is handled by github.Client that accepts BuildDetail)
- changed Commit map to use commit hashes as keys in testing.Repository since it works better with BuildDetail object

lynnsh commented on repository/client.go:
```diff
@@ -4,7 +4,8 @@ import "context"
+	GetCommit(ctx context.Context, commitURL string) (Commit, error)
 	GetOrganization(ctx context.Context, uri string) (Organization, error)
-	GetDefaultBranch(ctx context.Context, commitURL string) (DefaultBranch, error)
+	GetBranch(ctx context.Context, commit Commit) (Branch, error)
```
I think currently it gets the name of the Branch and BuildDetails: https://github.com/Shopify/voucher/pull/55/files#diff-ea35c13f3ad750a2fc1b8cc88d10a00dR9, this is the older file.

lynnsh commented on checks/org/check.go:
```diff
@@ -0,0 +1,50 @@
+	}
+
+	return true, nil
+}
```
Done.

lynnsh commented on cmd/config/checks.go:
```diff
@@ -82,6 +102,14 @@ func NewCheckSuite(metadataClient voucher.MetadataClient, names ...string) (*vou
+	//creating testing RepositoryClient for now (error is always nil for testing, so it is ignored)
+	repositoryClient, _ := vtesting.NewClient()
+	org := repository.Organization{
+		ID:   "MDEyOk9yZ2FuaXphdGlvbjgwODU=",
```
It was decided that ID is closely related to GitHub, so it was removed from repository.Organization.

catherinejones commented on cmd/config/checks.go:
```diff
@@ -82,6 +93,25 @@ func NewCheckSuite(metadataClient voucher.MetadataClient, names ...string) (*vou
+	//creating testing RepositoryClient for now (error is always nil for testing, so it is ignored)
+	repositoryClient, _ := vtesting.NewClient()
+
+	orgs := make(map[string]repository.Organization)
```
Could we throw the orgs processing logic into a separate file and function?

lynnsh replied:
Done.

catherinejones commented on cmd/config/checks.go:
```diff
@@ -82,6 +93,25 @@ func NewCheckSuite(metadataClient voucher.MetadataClient, names ...string) (*vou
+	orgs := make(map[string]repository.Organization)
+	repositories, ok := viper.Get("repositories").([]interface{})
+	if !ok {
+		return nil, fmt.Errorf("incorrectly configured repositories, failed to cast to []interface{}")
```
We shouldn't hard fail if we can't load the repositories, since we might not have defined them!

lynnsh replied:
Done (it returns an empty map if no repositories found).

catherinejones commented on repository/github/client.go:
```diff
@@ -41,38 +41,68 @@ func NewClient(ctx context.Context, auth *repository.Auth) (repository.Client, e
+		return repository.Commit{}, fmt.Errorf("Error getting repository metadata. Error: %s", err)
 	}
-	return commitInfo, nil
+	commitURI := fmt.Sprintf("https://github.com/%s/%s/commit/%s", repoMeta.Organization, repoMeta.Name, details.Commit)
```
Could we make this line it's own function that accepts a github.RepositoryMetadata? Or better yet, a `BuildDetail`!

lynnsh replied:
There was actually a function for this already, so I used it. I also created a similar function to get repositoryUrl.

catherinejones commented on checks/org/check.go:
```diff
@@ -0,0 +1,54 @@
+package org
```
Could you add a test to this package that tests this check with the `vtesting.RepositoryClient`?

lynnsh replied:
Done.

lynnsh commented on cmd/config/repo.go:
```diff
@@ -0,0 +1,30 @@
+	if strings.Contains(org.URL, "github.com") {
+		keyring, err := getRepositoryKeyRing()
+		if nil != err {
+			log.Errorf("failed to get github token: %s", err)
```
Done.

lynnsh commented on checks/org/check_test.go:
```diff
@@ -0,0 +1,82 @@
+	}
+	repoClient.AddRepository(r.CreateNewOrganization("Shopify", "https://github.com/Shopify"), "Repo_Shopify", commits, branches)
+
+	metadataClient, err := grafeas.NewClient(c, "shopify-docker-images", "shopify-binauthz", nil)
```
Done.

lynnsh commented on checks/org/check_test.go:
```diff
@@ -0,0 +1,82 @@
+	}
+	repoClient.AddRepository(r.CreateNewOrganization("Shopify", "https://github.com/Shopify"), "Repo_Shopify", commits, branches)
+
+	metadataClient, err := grafeas.NewClient(c, "voucher-test-project", "voucher-binauth", nil)
```
Done.

lynnsh commented on repository/github/client.go:
```diff
@@ -41,38 +41,69 @@ func NewClient(ctx context.Context, auth *repository.Auth) (repository.Client, e
-	if repoInfo.Resource.Commit.Repository.Owner.Typename != organizationType {
-		return repository.Organization{}, repository.NewTypeMismatchError(organizationType, repoInfo.Resource.Commit.Repository.Owner.Typename)
+
+	if repoInfo.Resource.Repository.Owner.Typename != organizationType {
```
Moved to newRepositoryOrgInfoResult.

lynnsh commented on repository/github/client.go:
```diff
@@ -41,38 +41,69 @@ func NewClient(ctx context.Context, auth *repository.Auth) (repository.Client, e
 	}
 
-	organization := repoInfo.Resource.Commit.Repository.Owner.Organization
+	organization := repoInfo.Resource.Repository.Owner.Organization
```
Done.

lynnsh commented on repository/github/client.go:
```diff
@@ -41,38 +41,69 @@ func NewClient(ctx context.Context, auth *repository.Auth) (repository.Client, e
+		return repository.Commit{}, fmt.Errorf("Error creating a commit url. Error: %s", err)
+	}
+
+	commit, err := newCommitInfoResult(ctx, ghc.ghClient, commitURI)
```
It is not [identical](https://developer.github.com/v4/object/commit/#fields), but it is based on it. I thought it only gets fields that are relevant later? I am not sure if there is a better way to reconstruct it?

lynnsh commented on repository/github/queries_test.go:
```diff
@@ -42,47 +42,45 @@ func (m *mockGitHubGraphQLClient) Query(ctx context.Context, query interface{},
-			testName:  "Testing happy path",
-			commitURL: "www.shopify.com",
+			testName: "Testing happy path",
+			uri:      "www.shopify.com",
```
This `uri` gets passed to newRepositoryOrgInfoResult: `newRepositoryOrgInfoResult(context.Background(), c, test.uri)`, which has a `url` variable used to query the organization. The url provided for testing should be repository url, so this was updated.

lynnsh commented on testing/metadata_client.go:
```diff
@@ -0,0 +1,66 @@
+	"github.com/Shopify/voucher/repository"
+)
+
+type MetadataClientMockup struct {
```
Done.

lynnsh commented on testing/metadata_client.go:
```diff
@@ -0,0 +1,66 @@
+)
+
+type MetadataClientMockup struct {
+	details []repository.BuildDetail
```
Done.

catherinejones commented:
When you have a moment could you run it through `golangci-lint` with the following options?

```shell
$ golangci-lint run --disable-all --enable=ineffassign --enable=misspell --enable=gocyclo --enable=vet --enable=gofmt --enable=golint
```

catherinejones commented:
Closing PR, as @lynnsh has split this into three separate PRs.
